lib/r10k/content_synchronizer.rb

```diff
@@ -8,31 +8,40 @@ def self.serial_accept(modules, visitor, loader)
     end
 
     def self.serial_sync(modules)
+      updated_modules = []
       modules.each do |mod|
-        mod.sync
+        updated = mod.sync
+        updated_modules << mod.name if updated
       end
+      updated_modules
     end
 
+    # Returns a Queue of the names of modules actually updated
     def self.concurrent_accept(modules, visitor, loader, pool_size, logger)
       mods_queue = modules_visit_queue(modules, visitor, loader)
       sync_queue(mods_queue, pool_size, logger)
     end
 
+    # Returns a Queue of the names of modules actually updated
     def self.concurrent_sync(modules, pool_size, logger)
       mods_queue = modules_sync_queue(modules)
       sync_queue(mods_queue, pool_size, logger)
     end
 
+    # Returns a Queue of the names of modules actually updated
     def self.sync_queue(mods_queue, pool_size, logger)
       logger.debug _("Updating modules with %{pool_size} threads") % {pool_size: pool_size}
-      thread_pool = pool_size.times.map { sync_thread(mods_queue, logger) }
+      updated_modules = Queue.new
+      thread_pool = pool_size.times.map { sync_thread(mods_queue, logger, updated_modules) }
       thread_exception = nil
 
       # If any threads raise an exception the deployment is considered a failure.
       # In that event clear the queue, wait for other threads to finish their
       # current work, then re-raise the first exception caught.
       begin
         thread_pool.each(&:join)
+        # Return the list of all modules that were actually updated
+        updated_modules
```

Comment:
In the block for `mods.each` there is the line `updated_modules << mod.name if udpated`; that's where `updated_modules` gets modified in a new thread. Does that answer your question?

Reply:
No. I'm not understanding how that modification of `updated_modules` persists outside of the `sync_thread` method, since `sync_thread` doesn't return the modified `updated_modules`, does it? All we do to `updated_modules` in this method is pass it as an argument to `sync_thread`, so I wouldn't expect its value to have changed by this point. But I probably just don't understand how threads work lol.

Reply:
Ok, I think maybe I see what you are asking about? The following code I think demonstrates what you are thinking of:

```
[143] pry(main)> def coolmethod(a, c); Thread.new {; while b = a.pop {; c <<  b ; }; end; } end
=> :coolmethod
[144] pry(main)> a = [1,2,3]
=> [1, 2, 3]
[145] pry(main)> c = []
=> []
[146] pry(main)> coolmethod(a,c)
=> #<Thread:0x00007fbbdb0ba830 (pry):86 dead>
[147] pry(main)> a
=> []
[148] pry(main)> c
=> []
```
We can see that the `pop` method removes the elements from the array passed in, but the `<<` does not successfully add elements do the second array. I think that magisus was expecting that second array to be populated...

Reply:
on the other hand, this PR does appear to work properly during my testing...

Reply:
huh, yeah, mine seems to have worked as Maggie intended:
```
irb(main):017:0> def thread_method(queue, empty_thing); Thread.new do; while element = queue.pop; empty_thing << element; end; end; end
=> :thread_method
irb(main):018:0> q = [1,2,3,4]
=> [1, 2, 3, 4]
irb(main):019:0> q
=> [1, 2, 3, 4]
irb(main):020:0> empty = []
=> []
irb(main):021:0> empty
=> []
irb(main):022:0> thread_method(q, empty)
=> #<Thread:0x00007fd19904c8c0@(irb):17 dead>
irb(main):023:0> q
=> []
irb(main):024:0> empty
=> [4, 3, 2, 1]
```